### PR TITLE
Allow for setting an explicit SpanContext on Span creation

### LINKED ISSUE #411: Allow SpanContext to be provided when starting a Span
I have a potential use-case where a process will be responsible for accepting incoming span payloads (could be in multiple wire formats) and then shipping them to various locations (initially Stackdriver) via one or more exporters.

Given that the incoming span payloads are fully formed, I'd like to be able to preserve the existing `SpanContext` when creating a new `Span` instance. The current implementation in `sdk/trace/span.go` always generates one for us:
https://github.com/open-telemetry/opentelemetry-go/blob/b80b336eb83d6765db23288d459c596669b10f02/sdk/trace/span.go#L255-L259

I'd like to propose allowing an explicit `SpanContext` to be set, using something like a `StartOption` (e.g. something like `func WithSpanContext(c SpanContext) StartOption`).

### PR DESCRIPTION
**Pull request by nicktrav**
In certain circumstances, information from an existing SpanContext is
desirable when instantiating a new Span. For instance, when a process is
accepting inbound spans in various formats before using an OpenTelemetry
exporter to forward to another destination.

Add a new StartOption, WithSpanContext, that allows setting an explicit
SpanContext on a trace.

Closes #411.

Signed-off-by: Nick Travers <n.e.travers@gmail.com>

### CODE REVIEW

File: sdk/trace/trace_test.go [hunk 1]
```diff
@@ -208,6 +208,28 @@ func TestSampling(t *testing.T) {
 	}
 }
 
+func TestStartSpanWithExplicitSpanContext(t *testing.T) {
```
Comment (lizthegrey) [inline sdk/trace/trace_test.go]: Please add test for invalid explicit span context.

Comment (nicktrav) [reply sdk/trace/trace_test.go]: Done!

**Comment by nicktrav** [comment]
@jmacd - thanks for the feedback. Some thoughts ...

> I'm not sure the specification would agree with this change--at least I haven't seen anything about supporting span creation with a pre-determined context. If we agree with this (I'm not sure), it should go in the spec first.

Agree - it seems as though creating a span with an explicit context is currently underspecified. If anywhere, it could probably be added to the list of API parameters for Span creation:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#span-creation

Something along the lines of:

> - `SpanContext` - override the current context, if provided. Defaults to propagating an existing parent context, if present, and ultimately falls back on seeding a new context with a new trace and span ID.

Thoughts? I can open a PR on the specs repo and we can discuss there.

> On this note, see how in #381 I added an explicit WithParent(SpanContext) option for specifying the parent span. That is supported by the specification.

Just to confirm, are you talking about this section in the spec?

> The parent Span or parent Span context, and whether the new Span should be a root Span. API MAY also have an option for implicit parent context extraction from the current context as a default behavior.

I was also thinking that we could tack on a sentence here saying that the API "MAY" have an option to add an explicit context, but it feels like this would be a useful feature of the API and therefore would benefit from being a standalone "MUST".

**Comment by lizthegrey** [comment]
Withdrawing my approval as per @jmacd's comments

**Comment by MrAlias** [comment]
@nicktrav it would be a good idea to either propose this in gitter ([the specification channel](https://gitter.im/open-telemetry/opentelemetry-specification) would be my guess of a good place) or in an [OTEP](https://github.com/open-telemetry/oteps) Issue/PR. I think gitter would be the better first step. You'll be able to get more immediate feedback/questions related to the proposed change there.

I'm interested to hear more about your use case. It isn't immediately clear to me why you would build a span context in a process just to have Open Telemetry act as a pass-through-proxy. But, like I said, I'm interested to better understand your use case.

**Comment by jmacd** [comment]
This requires a specification change, so I'm closing it. I imagine strong resistance to this idea as well, because of the potential for mis-use, but I think it's a worthwhile discussion because we may be able to find a better way to achieve the stated goal.

**Pull request closed without merging**